components/ee/ws-scheduler/pkg/scaler/driver.go

```diff
@@ -0,0 +1,380 @@
+
+	MaxGhostWorkspaces int           `json:"maxGhostWorkspaces"`
+	ReactionDelay      util.Duration `json:"reactionDelay"`
+
+	Renewal struct {
+		Interval   util.Duration `json:"interval"`
+		Percentage int           `json:"percentage"`
+	} `json:"renewal"`
+}
+
+// NewWorkspaceManagerPrescaleDriver creates a new WorkspaceManagerPrescale
+func NewWorkspaceManagerPrescaleDriver(config WorkspaceManagerPrescaleDriverConfig, controller Controller) (*WorkspaceManagerPrescaleDriver, error) {
+	if config.GhostOwner == "" {
+		config.GhostOwner = DefaultGhostOwner
+	}
+	if config.Renewal.Percentage < 0 && 100 < config.Renewal.Percentage {
+		return nil, xerrors.Errorf("renewal.percentage must be between 0 and 100 (inclusive)")
+	}
+
+	conn, err := grpc.Dial(config.WorkspaceManagerAddr, grpc.WithInsecure())
+	if err != nil {
+		return nil, xerrors.Errorf("cannot connect to ws-manager: %w", err)
+	}
+	return &WorkspaceManagerPrescaleDriver{
+		Config:     config,
+		Client:     api.NewWorkspaceManagerClient(conn),
+		conn:       conn,
+		Controller: controller,
+		stop:       make(chan struct{}),
+	}, nil
+}
+
+// WorkspaceManagerPrescaleDriver implements a prescale driver using ws-manager's ghost pods
+type WorkspaceManagerPrescaleDriver struct {
+	Config WorkspaceManagerPrescaleDriverConfig
+
+	Client api.WorkspaceManagerClient
+	conn   *grpc.ClientConn
+
+	Controller Controller
+
+	stop chan struct{}
+	once sync.Once
+}
+
+type workspaceStatus struct {
+	Count              WorkspaceCount
+	DeletionCandidates []string
+}
+
+// Run runs the prescale driver until Stop() is called
+func (wspd *WorkspaceManagerPrescaleDriver) Run() {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	statusChan := make(chan workspaceStatus)
+	go func() {
+		for {
+			err := wspd.maintainWorkspaceStatus(ctx, statusChan)
+			if err == context.Canceled {
+				return
+			}
+			if err != nil {
+				log.WithError(err).Error("cannot maintain workspace count")
+				time.Sleep(1 * time.Second)
+			}
+		}
+	}()
+
+	var renewal <-chan time.Time
+	if wspd.Config.Renewal.Percentage > 0 && wspd.Config.Renewal.Interval > 0 {
+		t := time.NewTicker(time.Duration(wspd.Config.Renewal.Interval))
+		defer t.Stop()
+		renewal = t.C
+	}
+	houseKeeping := time.NewTicker(1 * time.Minute)
+	defer houseKeeping.Stop()
+
+	var (
+		counts         = make(chan WorkspaceCount)
+		status         workspaceStatus
+		startingGhosts = make(map[string]time.Time)
+	)
+
+	delta := wspd.Controller.Control(ctx, counts)
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case status = <-statusChan:
+			for _, id := range status.DeletionCandidates {
+				delete(startingGhosts, id)
+			}
+			status.Count.Ghost += len(startingGhosts)
+			counts <- status.Count
+			log.WithField("counts", counts).Debug("status update")
+		case <-houseKeeping.C:
+			for id, t1 := range startingGhosts {
+				if time.Since(t1) < maxGhostStartTime {
```

Comment:
Why is this necessary? Are there ghosts taking too long to start? What are possible reasons?

Reply:
It's more of a fallback. I wanted to avoid the situation where we're accumulating ghosts that don't actually exist.
This case should not happen, as we're always listening to updates/receive a full status list when we reconnect.
I'll put a warning there, so in case it it does happen we'll learn about it.

Reply:
Hm. I think what I'm stumbling about is the fact that:
 - we have the controller output a `delta` instead of an absolute target value, which makes it more likely to change over time
 - we even increase the gap between calculation and execution